ddtrace/tracer/payload_v1.go

```diff
@@ -19,135 +32,1378 @@ package tracer
+		p.header[3] = msgpackMap32
+		p.readOff = 3
+	}
 }
 
-// Close implements io.Closer
 func (p *payloadV1) Close() error {
-	panic("not implemented")
+	p.clear()
+	return nil
 }
 
-// Write implements io.Writer. It writes data directly to the buffer.
-func (p *payloadV1) Write(data []byte) (n int, err error) {
-	panic("not implemented")
+func (p *payloadV1) Write(b []byte) (int, error) {
+	p.buf = append(p.buf, b...)
+	return len(b), nil
 }
 
 // Read implements io.Reader. It reads from the msgpack-encoded stream.
 func (p *payloadV1) Read(b []byte) (n int, err error) {
-	panic("not implemented")
+	if len(p.header) == 0 {
+		p.header = make([]byte, 8)
+		p.updateHeader()
+	}
+	if p.readOff < len(p.header) {
+		// reading header
+		n = copy(b, p.header[p.readOff:])
+		p.readOff += n
+		return n, nil
+	}
+	if len(p.buf) == 0 {
+		p.encode()
+	}
+	if p.reader == nil {
+		p.reader = bytes.NewReader(p.buf)
+	}
+	return p.reader.Read(b)
+}
+
+// encode writes existing payload fields into the buffer in msgp format.
+func (p *payloadV1) encode() {
+	st := newStringTable()
+	p.buf = encodeField(p.buf, p.bm, 2, p.containerID, st)
+	p.buf = encodeField(p.buf, p.bm, 3, p.languageName, st)
+	p.buf = encodeField(p.buf, p.bm, 4, p.languageVersion, st)
+	p.buf = encodeField(p.buf, p.bm, 5, p.tracerVersion, st)
+	p.buf = encodeField(p.buf, p.bm, 6, p.runtimeID, st)
+	p.buf = encodeField(p.buf, p.bm, 7, p.env, st)
+	p.buf = encodeField(p.buf, p.bm, 8, p.hostname, st)
+	p.buf = encodeField(p.buf, p.bm, 9, p.appVersion, st)
+
+	p.encodeAttributes(p.bm, 10, p.attributes, st)
+
+	p.encodeTraceChunks(p.bm, 11, p.chunks, st)
+}
+
+type fieldValue interface {
+	bool | []byte | int32 | int64 | uint32 | uint64 | string
+}
+
+// encodeField takes a field of any value and encodes it into the given buffer
+// in msgp format.
+func encodeField[F fieldValue](buf []byte, bm bitmap, fieldID uint32, a F, st *stringTable) []byte {
+	if !bm.contains(fieldID) {
+		return buf
+	}
+	buf = msgp.AppendUint32(buf, uint32(fieldID)) // msgp key
+	switch value := any(a).(type) {
+	case string:
+		// encode msgp value, either by pulling from string table or writing it directly
+		buf = st.Serialize(value, buf)
+	case bool:
+		buf = msgp.AppendBool(buf, value)
+	case float64:
+		buf = msgp.AppendFloat64(buf, value)
+	case int32, int64:
+		buf = msgp.AppendInt64(buf, handleIntValue(value))
+	case uint32:
+		buf = msgp.AppendUint32(buf, value)
+	case uint64:
+		buf = msgp.AppendUint64(buf, value)
+	case []byte:
+		buf = msgp.AppendBytes(buf, value)
+	case arrayValue:
+		buf = msgp.AppendArrayHeader(buf, uint32(len(value)))
+		for _, v := range value {
+			buf = v.encode(buf, st)
+		}
+	}
+	return buf
+}
+
+// encodeAttributes encodes an array associated with fieldID into the buffer in msgp format.
+// Each attribute is encoded as three values: the key, value type, and value.
+func (p *payloadV1) encodeAttributes(bm bitmap, fieldID int, kv map[string]anyValue, st *stringTable) (bool, error) {
```

Comment:
I do not see any reference to process tags being encoded in attributes though, did I miss a place where it's called or are our system-tests not catching the missing tags?

Reply:
Process tags added in [e759d0e](https://github.com/DataDog/dd-trace-go/pull/3947/commits/e759d0eb2017365d4644034bfd8f0b8da36b6df0)!